File: WOT-iOS/WOT-iOS/core/CustomDatasources/Menu/WOTMenuDatasource.m
//
//  WOTMenuDatasource.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 6/8/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTMenuDatasource.h"
#import "UserSession.h"
#import "WOTSessionDataProvider.h"

#import "WOTProfileViewController.h"
#import "WOTTankListViewController.h"
#import "WOTPlayersListViewController.h"

@interface WOTMenuDatasource () <NSFetchedResultsControllerDelegate>

@property (nonatomic, strong) NSMutableArray *availableViewControllers;
@property (nonatomic, strong) NSArray *visibleViewControllers;
@property (nonatomic, strong) NSFetchedResultsController *fetchedResultController;

@end

@implementation WOTMenuDatasource

- (id)init {

    self = [super init];
    if (self){
        

        self.availableViewControllers = [[NSMutableArray alloc] init];
        [self.availableViewControllers addObject:[[WOTMenuItem alloc] initWithClass:[WOTTankListViewController class] title:WOTString(WOT_STRING_TANKOPEDIA) image:nil userDependence:NO]];
        [self.availableViewControllers addObject:[[WOTMenuItem alloc] initWithClass:[WOTPlayersListViewController class] title:WOTString(WOT_STRING_PLAYERS) image:nil userDependence:NO]];
        [self.availableViewControllers addObject:[[WOTMenuItem alloc] initWithClass:[WOTProfileViewController class] title:WOTString(WOT_STRING_PROFILE) image:nil userDependence:YES]];

    
        NSManagedObjectContext *context = [[WOTCoreDataProvider sharedInstance] mainManagedObjectContext];
        NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:NSStringFromClass([UserSession class])];
        [fetchRequest setSortDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:WOT_KEY_EXPIRES_AT ascending:NO]]];
        self.fetchedResultController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest managedObjectContext:context sectionNameKeyPath:nil cacheName:nil];
        self.fetchedResultController.delegate = self;
        
        NSError *error = nil;
        [self.fetchedResultController performFetch:&error];
    
    }
    return self;
}

- (void)dealloc {

    self.fetchedResultController.delegate = nil;
}

- (void)rebuild {
    
    BOOL sessionHasBeenExpired = [WOTSessionDataProvider  sessionHasBeenExpired];
    if(sessionHasBeenExpired) {
        
        NSPredicate *predicate = nil;
        predicate = [NSPredicate predicateWithFormat:@"SELF.userDependence == NO"];
        self.visibleViewControllers = [self.availableViewControllers filteredArrayUsingPredicate:predicate];
    } else {
        
        self.visibleViewControllers = [self.availableViewControllers copy];
    }
}

- (void)setVisibleViewControllers:(NSArray *)visibleViewControllers {
    
    _visibleViewControllers = visibleViewControllers;
    [self.delegate hasUpdatedData:self];
}

- (NSInteger)objectsCount {
    
    return [self.visibleViewControllers count];
}

- (WOTMenuItem *)objectAtIndex:(NSInteger)index {
    
    return (WOTMenuItem *)(self.visibleViewControllers[index]);
}

#pragma mark - NSFetchedResultsControllerDelegate
- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    
    [self rebuild];

}
@end

File: WOT-iOS/WOT-iOS/Sources/WOTMenuViewController/datasource/WOTMenuDatasource.swift
//
//  WOTMenuDatasource.swift
//  WOT-iOS
//
//  Created on 8/14/18.
//  Copyright © 2018. All rights reserved.
//

import WOTApi
import WOTPivot

@objc
protocol WOTMenuDatasourceDelegate: NSObjectProtocol {
    func hasUpdatedData(_ datasource: WOTMenuDatasourceProtocol)
}

@objc
protocol WOTMenuDatasourceProtocol: NSObjectProtocol {
    var delegate: WOTMenuDatasourceDelegate? { get set }
    func object(at index: Int) -> WOTMenuItem?
    func objectsCount() -> Int
    func rebuild()
}

@objc
class WOTMenuDatasource: NSObject, WOTMenuDatasourceProtocol {
    // MARK: - Properties

    weak var delegate: WOTMenuDatasourceDelegate?

    var availableViewControllers: [WOTMenuItem] = []
    var visibleViewControllers: [WOTMenuItem]? {
        didSet {
            delegate?.hasUpdatedData(self)
        }
    }

    var fetchedResultController: NSFetchedResultsController<NSFetchRequestResult>?

    override init() {
        super.init()

        availableViewControllers.append(WOTMenuItem(controllerClass: WOTTankPivotViewController.self, controllerTitle: WOTApi.L10n.wotStringTankdeleyev, icon: UIImage(), userDependence: false))
        availableViewControllers.append(WOTMenuItem(controllerClass: WOTTankListViewController.self, controllerTitle: WOTApi.L10n.wotStringTankopedia, icon: UIImage(), userDependence: false))
        availableViewControllers.append(WOTMenuItem(controllerClass: WOTPlayersListViewController.self, controllerTitle: WOTApi.L10n.wotStringPlayers, icon: UIImage(), userDependence: false))
        availableViewControllers.append(WOTMenuItem(controllerClass: WOTProfileViewController.self, controllerTitle: WOTApi.L10n.wotStringProfile, icon: UIImage(), userDependence: false))
    }

    func object(at index: Int) -> WOTMenuItem? {
        return visibleViewControllers?[index]
    }

    func objectsCount() -> Int {
        return visibleViewControllers?.count ?? 0
    }

    func rebuild() {
        if WOTSessionManager.sessionHasBeenExpired() {
            let predicate = NSPredicate(format: "SELF.userDependence = NO")
            visibleViewControllers = availableViewControllers.filter { predicate.evaluate(with: $0) }
        } else {
            var visibleViewControllers = [WOTMenuItem]()
            visibleViewControllers.append(contentsOf: availableViewControllers)
            self.visibleViewControllers = visibleViewControllers
        }
    }
}

extension WOTMenuDatasource: NSFetchedResultsControllerDelegate {
    func controllerDidChangeContent(_: NSFetchedResultsController<NSFetchRequestResult>) {
        rebuild()
    }
}
